src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2362,37 +2362,73 @@ public override BoundNode VisitConditionalAccess(BoundConditionalAccess node)
 
         public override BoundNode VisitConditionalOperator(BoundConditionalOperator node)
         {
-            var isByRef = node.IsRef;
-
             VisitCondition(node.Condition);
             var consequenceState = this.StateWhenTrue;
             var alternativeState = this.StateWhenFalse;
+            bool consequenceReachable = this.StateWhenTrue.Reachable;
+            bool alternativeReachable = this.StateWhenFalse.Reachable;
+
+            TypeWithState consequenceRValue;
+            TypeWithState alternativeRValue;
+
+            if (node.IsRef)
+            {
+                TypeWithAnnotations consequenceLValue;
+                TypeWithAnnotations alternativeLValue;
+                (consequenceLValue, consequenceRValue) = visitConditionalRefOperand(consequenceState, node.Consequence);
+                consequenceState = this.State;
+                (alternativeLValue, alternativeRValue) = visitConditionalRefOperand(alternativeState, node.Alternative);
+                Join(ref this.State, ref consequenceState);
+
+                TypeSymbol refResultType = consequenceRValue.Type;
+                if (IsNullabilityMismatch(consequenceLValue, alternativeLValue))
+                {
```

Comment:
I think this should be a `type.MergeNullability` operation, followed by checking the conversions (which is where the errors would arise).

Reply:
`MergeNullability` always succeeds, so I don't see how that would work. 
`IsNullabilityMismatch` is what we use in other scenarios for checking `ref` compatibility.

---
In reply to: [268377530](https://github.com/dotnet/roslyn/pull/34323#discussion_r268377530) [](ancestors = 268377530)

Reply:
Yes, but you need that in order to compute the result nullability.  Below you simply take `consequenceRValue.Type` which is certainly not correct if only the alternative is reachable.

I don't think the reachability should affect the computed nested annotations of the result, but this is something that the LDM can confirm.  It is certainly the case that for regular type analysis; reachability never has any impact on the computed type (only the flow state).

In short, I think you need to
(1) compute the result `TypeSymbol`, which is composed of a type from merging the type from both sides (unconditionally)
(2) compute the result `NullableAnnotation` (either as below or simply `consequenceLValue.NullableAnnotation.EnsureCompatible(alternativeLValue.NullableAnnotation` but I think unconditionally is correct)
(3) check each reachable consequence or alternative against that resulting `TypeWithAnnotation` using `IsNullabilityMismatch`


---
In reply to: [268377989](https://github.com/dotnet/roslyn/pull/34323#discussion_r268377989) [](ancestors = 268377989,268377530)

Reply:
We can merge types unconditionally, as the types might be different.
I'll adjust the logic to not account for reachability to determine the result type.

Reply:
If the types are different then it is simply a hard error in semantics and you can use an error type.  If it is not a hard error we need to merge the types.  The types we produce by merging can be a type that did not appear in source (due to combining nullability annotations) and that is by design.

---
In reply to: [268450067](https://github.com/dotnet/roslyn/pull/34323#discussion_r268450067) [](ancestors = 268450067)

Reply:
It is fine to produce a type that doesn't appear in source. What I'm objecting to is providing error messages that unnecessarily reference such a type (ie. is confusing to users). 
Also, as I've illustrated with `ConditionalOperator_Ref_NestedNullabilityMismatch`, doing two mismatch checks would mean showing two indirect diagnostics (or picking one of the two if we made the second check conditional) when a single check suffices.